### ISSUE TITLE
Race condition in high-level API's may leave behind  .fuse_hiddenXXX files

### ISSUE THREAD

**Issue by devPeri** [libfuse/libfuse#746]
# Issue Description
When an open file is closed and immediately deleted and the entry is cached, FUSE_RELEASE and FUSE_UNLINK are sent back to back. 

Handling FUSE_RELEASE, `fuse_do_release()` calls `fuse_fs_release()` outside the lock.

If `fuse_fs_release()` is slow to complete, the FUSE_UNLINK will arrive and start processing in another thread.
`fuse_lib_unlink()` will see that the file is open and call `fuse_fs_rename()` via `hide_node()` to rename the file to `.fuse_hiddenXXX`.

`fuse_fs_rename()` is also called outside the lock and while it is processing, the `fuse_fs_release()` has completed and `fuse_do_release()` acquires the lock and continues processing.

As `fuse_fs_rename()` is still processing, `rename_node()` has not been called yet and thus `node->is_hidden` is still `0`.

With `fuse_do_release()` now processing with the lock, `node->open_count` is decremented to `0` but because `node->is_hidden` is still `0`  the `unlink_hidden` flag is never set to `1` and thus the file is considered as closed but not deleted.

`fuse_do_release()` has now finished completely, and `fuse_lib_unlink()` is still waiting for `fuse_fs_rename()` to complete.
Once `fuse_fs_rename()` completes, the processing resumes and `rename_node()` is called to hide the node etc.

**But by this time the file has already been closed in the user-space and the file has been renamed to `.fuse_hiddenXXX`.
As the file is already closed, `fuse_do_release()` will not be called to clean-up the `.fuse_hiddenXXX` file.**

If entry caching is not used, between FUSE_RELEASE and FUSE_UNLINK there is a whole bunch of FUSE_GETATTR happening that allow FUSE_RELEASE to complete processing before FUSE_UNLINK starts.

## Code flaw
```
static void fuse_do_release(struct fuse *f, fuse_ino_t ino, const char *path,
                            struct fuse_file_info *fi)
{
        struct node *node;
        int unlink_hidden = 0;

        fuse_fs_release(f->fs, path, fi);        <--- called here without any entry lock protection

        pthread_mutex_lock(&f->lock);
        ...
```

# Potential fix
**Call fuse_fs_release() after the lock is acquired.**
I don't know if this is the best solution or if there is reason why it should not be called with the lock held.

# Sample debug logs
With my setup using a custom FUSE based FS, I can easily reproduce this by simply using `vim` to edit a file (`vim` creates and deletes the `.swp` and `.swpx` files)

## Example RELEASE and UNLINK overlap
Using libfuse origin/master.

This example includes custom debug outputs from my own filesystem implementation (lines starting with timestamps e.g. "2023-03-16|14:42:29.191551|")

Also added additional debugging within libfuse (`is_open[]` and `do_release[]` lines) to show the states of `open_count`, `is_hidden` and `unlink_hidden`.

```
unique: 44, opcode: LOOKUP (1), nodeid: 4, insize: 53, pid: 251520
LOOKUP /FS/00000000000000000000000000000000/25/.asdasd.swpx
getattr[NULL] /FS/00000000000000000000000000000000/25/.asdasd.swpx
2023-03-16|14:42:29.191374| getattr: /FS/00000000000000000000000000000000/25/.asdasd.swpx stbuf:0x7f4b59ffac90 fi:(nil)
2023-03-16|14:42:29.191493| getattr: /FS/00000000000000000000000000000000/25/.asdasd.swpx csrbVFS::getattr: ret:5
   unique: 44, error: -2 (No such file or directory), outsize: 16

unique: 46, opcode: LOOKUP (1), nodeid: 4, insize: 53, pid: 251520
LOOKUP /FS/00000000000000000000000000000000/25/.asdasd.swpx
getattr[NULL] /FS/00000000000000000000000000000000/25/.asdasd.swpx
2023-03-16|14:42:29.191551| getattr: /FS/00000000000000000000000000000000/25/.asdasd.swpx stbuf:0x7f4b597f9c90 fi:(nil)
2023-03-16|14:42:29.191644| getattr: /FS/00000000000000000000000000000000/25/.asdasd.swpx csrbVFS::getattr: ret:5
   unique: 46, error: -2 (No such file or directory), outsize: 16

unique: 48, opcode: CREATE (35), nodeid: 4, insize: 69, pid: 251520
create flags: 0x80c2 /FS/00000000000000000000000000000000/25/.asdasd.swpx 0100600 umask=0022
2023-03-16|14:42:29.191690| create: /FS/00000000000000000000000000000000/25/.asdasd.swpx mode:100600 fi:0x7f4b597f9d30 fi->flags:100302
2023-03-16|14:42:29.194106| create: /FS/00000000000000000000000000000000/25/.asdasd.swpx mode:100600 fi:0x7f4b597f9d30 fi->fh:7f4b18002ac0 ret:555
   create[139960501938880] flags: 0x80c2 /FS/00000000000000000000000000000000/25/.asdasd.swpx
getattr[139960501] /FS/00000000000000000000000000000000/25/.asdasd.swpx
2023-03-16|14:42:29.194130| getattr: /FS/00000000000000000000000000000000/25/.asdasd.swpx stbuf:0x7f4b597f9c50 fi:0x7f4b597f9d30
2023-03-16|14:42:29.194202| getattr: /FS/00000000000000000000000000000000/25/.asdasd.swpx csrbVFS::getattr: ret:555
2023-03-16|14:42:29.194214| getattr: /FS/00000000000000000000000000000000/25/.asdasd.swpx return mode:100600 size:0
   NODEID: 6
   unique: 48, success, outsize: 160

unique: 50, opcode: FLUSH (25), nodeid: 6, insize: 64, pid: 251520
flush[139960501938880]
2023-03-16|14:42:29.194269| flush: (null) fi:0x7f4b59ffad30 fi->fh:7f4b18002ac0
2023-03-16|14:42:29.194294| flush: (null) fi:0x7f4b59ffad30 fi->fh:7f4b18002ac0 ret:555
lock[139960501938880] F_SETLK F_UNLCK start: 0 len: 0 pid: 0
2023-03-16|14:42:29.194306| lock: (null) fi:0x7f4b59ffad30 cmd:6 fl:0x7f4b59ffac20 lock_owner:b5254441f7b59131
2023-03-16|14:42:29.194316| lock: (null) F_SETLK l_type:2
2023-03-16|14:42:29.194362| lock: (null) ret:555
   unique: 50, success, outsize: 16

unique: 54, opcode: GETATTR (3), nodeid: 4, insize: 56, pid: 251520
getattr[NULL] /FS/00000000000000000000000000000000/25
2023-03-16|14:42:29.194397| getattr: /FS/00000000000000000000000000000000/25 stbuf:0x7f4b59ffac60 fi:(nil)
2023-03-16|14:42:29.194455| getattr: /FS/00000000000000000000000000000000/25 csrbVFS::getattr: ret:555
2023-03-16|14:42:29.194461| getattr: /FS/00000000000000000000000000000000/25 return mode:40755 size:32
   unique: 54, success, outsize: 120

unique: 52, opcode: RELEASE (18), nodeid: 6, insize: 64, pid: 0
unique: 56, opcode: UNLINK (10), nodeid: 4, insize: 53, pid: 251520
is_open[.asdasd.swpx] open:1 node:0x7f4bb400d388 open_count:1
release[139960501938880] flags: 0x8002
getattr[NULL] /FS/00000000000000000000000000000000/25/.fuse_hidden0000000600000001
2023-03-16|14:42:29.194529| release: (null) fi:0x7f4b597f9d30 fi->fh:7f4b18002ac0
2023-03-16|14:42:29.194531| getattr: /FS/00000000000000000000000000000000/25/.fuse_hidden0000000600000001 stbuf:0x7f4b59ffac10 fi:(nil)
2023-03-16|14:42:29.194606| getattr: /FS/00000000000000000000000000000000/25/.fuse_hidden0000000600000001 csrbVFS::getattr: ret:5
rename /FS/00000000000000000000000000000000/25/.asdasd.swpx /FS/00000000000000000000000000000000/25/.fuse_hidden0000000600000001 0x0
2023-03-16|14:42:29.194618| rename: /FS/00000000000000000000000000000000/25/.asdasd.swpx -> /FS/00000000000000000000000000000000/25/.fuse_hidden0000000600000001 flags:0
2023-03-16|14:42:29.194796| release: (null) fi:0x7f4b597f9d30 fi->fh:0 ret:555
do_release[(null)]-A open_count:1 is_hidden:0
do_release[(null)]-B open_count:0 is_hidden:0 unlink_hidden:0
   unique: 52, success, outsize: 16
2023-03-16|14:42:29.196305| rename: /FS/00000000000000000000000000000000/25/.asdasd.swpx -> /FS/00000000000000000000000000000000/25/.fuse_hidden0000000600000001 ret:555
   unique: 56, success, outsize: 16
unique: 60, opcode: FLUSH (25), nodeid: 5, insize: 64, pid: 251520
flush[139960703297072]
unique: 58, opcode: FORGET (2), nodeid: 6, insize: 48, pid: 0
```

## Another example RELEASE and UNLINK overlap
Using debian testing:
```
Package: libfuse3-3                      
Version: 3.14.0-2
```
Includes some manual annotations.

```
unique: 68, opcode: FLUSH (25), nodeid: 7, insize: 64, pid: 258058
flush[.asdasd.swp]
2023-03-16|15:27:24.880763| flush: (null) fi:0x7f40817f9d10 fi->fh:7f406c002f20
2023-03-16|15:27:24.880791| flush: (null) fi:0x7f40817f9d10 fi->fh:7f406c002f20 ret:555
lock[.asdasd.swp] F_SETLK F_UNLCK start: 0 len: 0 pid: 0
2023-03-16|15:27:24.880803| lock: (null) fi:0x7f40817f9d10 cmd:6 fl:0x7f40817f9bf0 lock_owner:d906932364f1173f
2023-03-16|15:27:24.880810| lock: (null) F_SETLK l_type:2
2023-03-16|15:27:24.880912| lock: (null) ret:555
   unique: 68, success, outsize: 16

*** RELEASE STARTING
unique: 70, opcode: RELEASE (18), nodeid: 7, insize: 64, pid: 0
release[.asdasd.swp] flags: 0x8002
2023-03-16|15:27:24.880941| release: (null) fi:0x7f40817f9d10 fi->fh:7f406c002f20

unique: 72, opcode: GETATTR (3), nodeid: 6, insize: 56, pid: 258058
getattr[NULL] /FS/00000000000000000000000000000000/27
2023-03-16|15:27:24.880960| getattr: /FS/00000000000000000000000000000000/27 stbuf:0x7f40827fbc30 fi:(nil)
2023-03-16|15:27:24.881084| getattr: /FS/00000000000000000000000000000000/27 csrbVFS::getattr: ret:555
2023-03-16|15:27:24.881093| getattr: /FS/00000000000000000000000000000000/27 return mode:40755 size:16
   unique: 72, success, outsize: 120

*** UNLINK STARTING
unique: 74, opcode: UNLINK (10), nodeid: 6, insize: 52, pid: 258058
getattr[NULL] /FS/00000000000000000000000000000000/27/.fuse_hidden0000000700000001
2023-03-16|15:27:24.881126| getattr: /FS/00000000000000000000000000000000/27/.fuse_hidden0000000700000001 stbuf:0x7f40827fbb90 fi:(nil)
2023-03-16|15:27:24.881167| release: (null) fi:0x7f40817f9d10 fi->fh:0 ret:555
   unique: 70, success, outsize: 16
*** RELEASE FINISHED

2023-03-16|15:27:24.881203| getattr: /FS/00000000000000000000000000000000/27/.fuse_hidden0000000700000001 csrbVFS::getattr: ret:5
rename /FS/00000000000000000000000000000000/27/.asdasd.swp /FS/00000000000000000000000000000000/27/.fuse_hidden0000000700000001 0x0
2023-03-16|15:27:24.881235| rename: /FS/00000000000000000000000000000000/27/.asdasd.swp -> /FS/00000000000000000000000000000000/27/.fuse_hidden0000000700000001 flags:0
2023-03-16|15:27:24.882792| rename: /FS/00000000000000000000000000000000/27/.asdasd.swp -> /FS/00000000000000000000000000000000/27/.fuse_hidden0000000700000001 ret:555
   unique: 74, success, outsize: 16
*** UNLINK FINISHED
```

## RELEASE and UNLINK processing with no entry caching
RELEASE has enough time to complete while a bunch of GETATTRs are executed before UNLINK is requested.

```
unique: 280, opcode: FLUSH (25), nodeid: 5, insize: 64, pid: 258473
flush[.asdasd.swp]
2023-03-16|15:29:00.480551| flush: (null) fi:0x7febaf7fdd10 fi->fh:7feb90001fc0
2023-03-16|15:29:00.480556| flush: (null) fi:0x7febaf7fdd10 fi->fh:7feb90001fc0 ret:555
lock[.asdasd.swp] F_SETLK F_UNLCK start: 0 len: 0 pid: 0 
2023-03-16|15:29:00.480560| lock: (null) fi:0x7febaf7fdd10 cmd:6 fl:0x7febaf7fdbf0 lock_owner:d9d5d502a9cf36b7
2023-03-16|15:29:00.480563| lock: (null) F_SETLK l_type:2
2023-03-16|15:29:00.480581| lock: (null) ret:555
   unique: 280, success, outsize: 16

*** RELEASE STARTING
unique: 282, opcode: RELEASE (18), nodeid: 5, insize: 64, pid: 0
release[.asdasd.swp] flags: 0x8002
unique: 284, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 258473
2023-03-16|15:29:00.480609| release: (null) fi:0x7febae7fbd10 fi->fh:7feb90001fc0
getattr[NULL] /
2023-03-16|15:29:00.480613| getattr: / stbuf:0x7febaeffcc30 fi:(nil)
2023-03-16|15:29:00.480616| getattr: / csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480619| getattr: / return mode:40777 size:0
   unique: 284, success, outsize: 120
unique: 286, opcode: LOOKUP (1), nodeid: 1, insize: 43, pid: 258473
LOOKUP /FS
getattr[NULL] /FS
2023-03-16|15:29:00.480632| getattr: /FS stbuf:0x7febaf7fdc60 fi:(nil)
2023-03-16|15:29:00.480639| getattr: /FS csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480642| getattr: /FS return mode:40777 size:0
   NODEID: 2
   unique: 286, success, outsize: 144
unique: 288, opcode: GETATTR (3), nodeid: 2, insize: 56, pid: 258473
getattr[NULL] /FS
2023-03-16|15:29:00.480653| getattr: /FS stbuf:0x7febaf7fdc30 fi:(nil)
2023-03-16|15:29:00.480659| getattr: /FS csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480662| getattr: /FS return mode:40777 size:0
   unique: 288, success, outsize: 120
unique: 290, opcode: LOOKUP (1), nodeid: 2, insize: 73, pid: 258473
LOOKUP /FS/00000000000000000000000000000000
getattr[NULL] /FS/00000000000000000000000000000000
2023-03-16|15:29:00.480674| getattr: /FS/00000000000000000000000000000000 stbuf:0x7febaeffcc60 fi:(nil)
2023-03-16|15:29:00.480693| release: (null) fi:0x7febae7fbd10 fi->fh:0 ret:555
   unique: 282, success, outsize: 16
*** RELEASE FINISHED

2023-03-16|15:29:00.480699| getattr: /FS/00000000000000000000000000000000 csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480702| getattr: /FS/00000000000000000000000000000000 return mode:40777 size:448
   NODEID: 3
   unique: 290, success, outsize: 144
unique: 292, opcode: GETATTR (3), nodeid: 3, insize: 56, pid: 258473
getattr[NULL] /FS/00000000000000000000000000000000 
2023-03-16|15:29:00.480718| getattr: /FS/00000000000000000000000000000000 stbuf:0x7febaf7fdc30 fi:(nil)
2023-03-16|15:29:00.480745| getattr: /FS/00000000000000000000000000000000 csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480753| getattr: /FS/00000000000000000000000000000000 return mode:40777 size:448
   unique: 292, success, outsize: 120
unique: 294, opcode: LOOKUP (1), nodeid: 3, insize: 43, pid: 258473
LOOKUP /FS/00000000000000000000000000000000/28
getattr[NULL] /FS/00000000000000000000000000000000/28
2023-03-16|15:29:00.480767| getattr: /FS/00000000000000000000000000000000/28 stbuf:0x7febae7fbc60 fi:(nil)
2023-03-16|15:29:00.480792| getattr: /FS/00000000000000000000000000000000/28 csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480794| getattr: /FS/00000000000000000000000000000000/28 return mode:40755 size:16
   NODEID: 4
   unique: 294, success, outsize: 144
unique: 296, opcode: GETATTR (3), nodeid: 4, insize: 56, pid: 258473
getattr[NULL] /FS/00000000000000000000000000000000/28
2023-03-16|15:29:00.480806| getattr: /FS/00000000000000000000000000000000/28 stbuf:0x7febaeffcc30 fi:(nil)
2023-03-16|15:29:00.480828| getattr: /FS/00000000000000000000000000000000/28 csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480831| getattr: /FS/00000000000000000000000000000000/28 return mode:40755 size:16
   unique: 296, success, outsize: 120
unique: 298, opcode: LOOKUP (1), nodeid: 4, insize: 52, pid: 258473
LOOKUP /FS/00000000000000000000000000000000/28/.asdasd.swp 
getattr[NULL] /FS/00000000000000000000000000000000/28/.asdasd.swp
2023-03-16|15:29:00.480843| getattr: /FS/00000000000000000000000000000000/28/.asdasd.swp stbuf:0x7febaf7fdc60 fi:(nil)
2023-03-16|15:29:00.480868| getattr: /FS/00000000000000000000000000000000/28/.asdasd.swp csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480871| getattr: /FS/00000000000000000000000000000000/28/.asdasd.swp return mode:100600 size:0
   NODEID: 5
   unique: 298, success, outsize: 144
unique: 300, opcode: GETATTR (3), nodeid: 4, insize: 56, pid: 258473
getattr[NULL] /FS/00000000000000000000000000000000/28
2023-03-16|15:29:00.480886| getattr: /FS/00000000000000000000000000000000/28 stbuf:0x7febae7fbc30 fi:(nil)
2023-03-16|15:29:00.480908| getattr: /FS/00000000000000000000000000000000/28 csrbVFS::getattr: ret:555
2023-03-16|15:29:00.480911| getattr: /FS/00000000000000000000000000000000/28 return mode:40755 size:16
   unique: 300, success, outsize: 120

*** UNLINK STARTING
unique: 302, opcode: UNLINK (10), nodeid: 4, insize: 52, pid: 258473
unlink /FS/00000000000000000000000000000000/28/.asdasd.swp
2023-03-16|15:29:00.480922| unlink: /FS/00000000000000000000000000000000/28/.asdasd.swp
2023-03-16|15:29:00.481293| unlink: /FS/00000000000000000000000000000000/28/.asdasd.swp ret:555
   unique: 302, success, outsize: 16    
*** UNLINK FINISHED
```

**Comment by Nikratio** [comment]
Thanks for the detailed report and investigation! 

I do not think I'll have time to look into this in the near future, so pull requests to address this are welcome.

**Comment by matthiasgoergens** [comment]
@devPeri Could you make the code that you used to generate those logs available?  Thanks!

I am trying to write a minimal test for this problem.

**Comment by matthiasgoergens** [comment]
Thanks!

Btw, I've only recently started contributing. You might want to give it a go, too. The people are friendly.

When implementing my own file systems, I found the low level API a lot better to work with.

**Comment by devPeri** [comment]
Found a way to replicate the issue with passthrough_fh. 
Slow down release() and rename()  ([passthrough_fh.txt](https://github.com/libfuse/libfuse/files/11221599/passthrough_fh.txt)) and just run vim to edit a file and you'll trigger the race condition between release() and unlink().

I tested this with Codespaces and I get it to trigger all the time.
Compile and run the patched passthrough_fh with the delays, and run e.g. "vim asd.txt" and just exist at a directory under the FUSE mountpoint, and you'll get the fuse_hidden leftovers.

I think I also have a proper patch that fixes this issue, by adding a bit more intelligence in unlink(), without affecting release() and the locks.
I'll test this a bit more and create a pull request for review.

**Comment by matthiasgoergens** [comment]
Awesome!

Btw, if you could automate the test (instead of eg running vim manually) that would be greatly appreciated.

That way we can add your test to the automated tests, and make sure that we don't accidentally reintroduce the bug next year.

**Comment by devPeri** [comment]
I think I figured out why this happens under specific conditions.

If RELEASE and UNLINK opcodes are sent back to back, and `fuse_fs_release()` and `fuse_fs_rename()` are slow to execute, UNLINK will run while RELEASE is still executing, UNLINK will try to rename the file, and while the rename is happening the RELEASE will finish executing.
So at the end, RELEASE will not detect in time that UNLINK has happened, and UNLINK will not detect in time that RELEASE has happened.

**Now, this is triggered only when `nullpath_ok` is set.**

If it is **not** SET then `get_path_nullok()` called by `fuse_lib_release()` will call `get_path_common()` and lock the path, and then the `fuse_lib_unlink()` will wait for the path to be unlocked before executing and thus synchronise with `fuse_lib_release()`.

If it **is** SET then `get_path_nullok()` will just set the path to null and return without locking anything and thus allowing `fuse_lib_unlink()` to eventually execute unimpeded while `fuse_lib_release()` is still running.

**Comment by devPeri** [comment]
Here is a one-liner test:
Start FUSE: `./passthrough_fh -f -d /mnt/temp`
Run test: ```bash -c 'T=`mktemp -d /mnt/temp/tmp/test746-XXXX` ; cd $T ; python3 -c "import os; f=open(\"test\", \"xt\"); f.close(); os.remove(\"test\");" ; ls -al'```

Here is a bit more elaborate custom [test](https://github.com/devPeri/libfuse/commit/dfb3ef57c32f2df3914aaaae0220663cf4da162d) for this issue from my branch, using `LD_PRELOAD` to inject delays in libc `release()` and `rename()`.
You can set the `TEST746_DELAY_DISABLE` env variable to disable the delay on demand at runtime.

**Comment by devPeri** [comment]
Fix in PR #782 for review.

**Comment by devPeri** [comment]
Fixed by #782
